### ISSUE TITLE
Incorrect folder structure with .net core on macOS

### ISSUE THREAD

**Issue by Psybernetic** [reinforced/Reinforced.Typings#71]
Hi,

I was using RT in an ASP.NET Core app (.NET Core SDK 2.1.x). I used Jetbrains Rider IDE and everything is running on macOS. 

My model is defined as:

```C#
using Reinforced.Typings.Attributes;

namespace MyAspNetApp.Models
{
    [TsClass]
    public class WeatherForecast
    {
        public string DateFormatted { get; set; }
        public int TemperatureC { get; set; }
        public string Summary { get; set; }

        public int TemperatureF
        {
            get { return 32 + (int) (TemperatureC / 0.5556); }
        }
    }
}
```

In the Reinforced.Typings.setting.xml file I specified the output directory as:

```XML
<RtTargetDirectory>$(ProjectDir)MyAspNetApp\src\app\_models</RtTargetDirectory>
```
However, the TypeScript files get put in a weird directory structure that looks like:

.
├── ...
├── src
│   ├── app
│   │   ├── MyAspNetApp/Models

And I was expecting:
.
├── ...
├── src
│   ├── app
│   │   ├── MyAspNetApp
│   │   │   ├── Models

In other words, it outputs to a folder called 'MyAspNetApp/Models' instead of creating a folder 'MyAspNetApp' folder and placing the files in a subdirectory called 'Models'.

I noticed this behaviour while using Jetbrains Rider and Visual Studio Code on macOS.

Any help would be appreciated.

Cheers,
Psybernetic

***P.S. It works perfectly when building with Visual Studio 2017 on Windows OS.***

**Comment by pavel-b-novikov** [comment]
Seems that .NET's ```Directory.CreateDirectory``` works weirdly on Max

**Comment by PandaWood** [comment]
On a Mac, the only character not allowed in file names is colon ":" - anything else is valid
On Windows, it bars on any of these  / \ : * ? " < > | 
So looks like you can't rely on just throwing a slash in a filename and hoping it will create a subdirectory because on a Mac it's a perfectly valid filename.

**Comment by PandaWood** [comment]
In the class `ReferenceInspector` method `GetPathForType` there's some code that although starts wisely and uses `Path.Combine` it then goes and replaces slashes with backslashes...
Which seems like the kind of naughty thing that might be going on here.

`Path.Combine(_context.TargetDirectory, fromConfiguration.Replace("/", "\\")).Replace("/", "\\")`

I actually get 10 out of the 60 tests failing on Mono - so I'm just looking at this now.
I think it's partly this and partly the use of "D:\\" that causes most of the issues.

**Comment by PandaWood** [comment]
FYI I am working on this out of interest...
And one other problem with the tests I found - that is more painful to fix - is some are CLR-specific and rely on the order of lists.

So, the TypeScript output produced in Mono (eg just use JetBrains Rider on a Mac) has lines in different order (but ultimately identical) and the tests (Tokenizer) fails expecting the lines to be in exactly the same order.

I suppose this could be a real issue, other than in testing, if multiple developers are on different CLR/platforms - the files produced by **Reinforced** would be "different" depending on platform and if they're in source control, would constantly show the files as "modified" even though they're actually not.

**Comment by pavel-b-novikov** [comment]
Sure!

**Comment by PandaWood** [comment]
Just one more comment on this, sorry
I am reproducing this using the **Samples** in the project rather than try to emulate the OP
So in [Reinforced.Settings.xml](https://github.com/reinforced/Reinforced.Typings/blob/master/Samples/Simple/Reinforced.Typings.Samples.Simple.Quickstart/Reinforced.Typings.settings.xml)

If I set `RtDivideTypesAmongFiles` to `true` and use the `<RtTargetDirectory>$(ProjectDir)Scripts\MyApplication\MyModel</RtTargetDirectory>`

Then the output is: 
![image](https://user-images.githubusercontent.com/41550/46273402-b2302280-c598-11e8-876c-45b929c3f869.png)

**Comment by PandaWood** [comment]
I just have one problem at the moment: **trying to use the code I've changed to test these samples**.

Unfortunately, the Samples solution imports the latest nuget packge of Reinforced.Typings instead of importing the sln on disk.
And because I'm a little bit clueless, I'm not sure how to get them to reference the actual Reinforced.Typings.sln (I think we should change it to do this if it's easy)

I tried importing/adding the dll/projects in to the Samples solution but couldn't it get to work (output typescript) on build.

**Comment by pavel-b-novikov** [comment]
Answered on twitter

**Comment by romansp** [comment]
I'm trying to setup build on MacOS, and I have following error occurring during build.

`Reinforced.Typings : IO error RT0001: Could not acuire temorary file \Users\roman.pavlov\Documents\dev\ts\Models.Shared\models.ts.tmp: Path cannot be the empty string or all whitespace. [/Users/roman.pavlov/Documents/dev/ts/Models.Shared/Models.Shared.csproj]`

<details>
<summary>Reinforced.Typings.settings.xml</summary>

```
<?xml version="1.0" encoding="utf-8"?>
<Project>
  <PropertyGroup>
    <RtTargetFile>$(ProjectDir)models.ts</RtTargetFile>
    <RtConfigurationMethod>Models.Shared.TypescriptGenerationConfiguration.Configure</RtConfigurationMethod>
    <RtDivideTypesAmongFiles>true</RtDivideTypesAmongFiles>
    <RtTargetDirectory>$(ProjectDir)</RtTargetDirectory>
    <RtBypassTypeScriptCompilation>false</RtBypassTypeScriptCompilation>
    <RtDisable>false</RtDisable>
  </PropertyGroup>
</Project>
```

</details>

Notice how full path to temporary file is constructed using backslashes and then it fails here: https://github.com/reinforced/Reinforced.Typings/blob/d11a460f79ca646ca20ee7ee03c3ff52b5c39d06/Reinforced.Typings/FilesOperations.cs#L86-L87

`Path.GetDirectoryName` returns empty string because constructed path makes no sense on forward-slash filesystem and then `Directory.Exists(dir)` call throws exception.

I think it's the same cause you already trying to solve in this issue. It will be huge :+1: if you can fix it and make it work cross-platform.

**Comment by PandaWood** [comment]
Hi @romansp yes, I believe I have this fixed, I am just doing some physical testing now before submit a pull request (I also fixed the 2 spelling mistakes in that error message in a separate PR ;-))

**Comment by PandaWood** [comment]
Yes, so in addition to getting the tests to pass, my physical test looks successful and I'll now prepare my branch for PR submission.

In the screenshot below you can see the old/incorrect filename with backslashes
And the new full file path below it, created an hour or so later after I changed the nuget package to use my fixes.

![image](https://user-images.githubusercontent.com/41550/46569924-cfe9f700-c99f-11e8-8ea7-3eee52424258.png)

**Comment by romansp** [comment]
This is great @PandaWood. 

Yeah, I noticed that another PR dealing with typos.Thank you. Don't hesitate if you need some extra help with testing it on MacOS.

**Comment by romansp** [comment]
It seems that fix was merged to master some time ago, thanks @PandaWood for the efforts. Is there anything that stops new version from being published to NuGet?

**Comment by pavel-b-novikov** [comment]
Absence of free time

**Comment by romansp** [comment]
Sure thing, no pressure at all 👍

**Comment by pavel-b-novikov** [comment]
Seems that original issue is fixed by @PandaWood in version 1.4.97.
Please, everyone, say thanks to @PandaWood.

**Comment by Psybernetic** [comment]
Thank you very much! @PandaWood
